nova_vm/src/ecmascript/builtins/indexed_collections/array_objects/array_iterator_objects/array_iterator_prototype.rs

```diff
@@ -0,0 +1,142 @@
+}
+
+impl ArrayIteratorPrototype {
+    fn next(agent: &mut Agent, this_value: Value, _arguments: ArgumentsList) -> JsResult<Value> {
+        // 27.5.3.2 GeneratorValidate ( generator, generatorBrand )
+        // 3. If generator.[[GeneratorBrand]] is not generatorBrand, throw a TypeError exception.
+        let Value::ArrayIterator(iterator) = this_value else {
+            return Err(agent.throw_exception_with_static_message(
+                ExceptionType::TypeError,
+                "ArrayIterator expected",
+            ));
+        };
+
+        // 23.1.5.1 CreateArrayIterator ( array, kind ), step 1. b
+        // NOTE: We set `array` to None when the generator in the spec text has returned.
+        let Some(array) = agent[iterator].array else {
+            return Ok(create_iter_result_object(agent, Value::Undefined, true).into_value());
+        };
+
+        let len: i64 = match array {
+            // i. If array has a [[TypedArrayName]] internal slot, then
+            // TODO!
+            Object::Int8Array(_)
+            | Object::Uint8Array(_)
+            | Object::Uint8ClampedArray(_)
+            | Object::Int16Array(_)
+            | Object::Uint16Array(_)
+            | Object::Int32Array(_)
+            | Object::Uint32Array(_)
+            | Object::BigInt64Array(_)
+            | Object::BigUint64Array(_)
+            | Object::Float32Array(_)
+            | Object::Float64Array(_) => todo!(),
+            // ii. Else,
+            //     1. Let len be ? LengthOfArrayLike(array).
+            Object::Array(array) if agent[array].object_index.is_none() => array.len(agent).into(),
+            _ => length_of_array_like(agent, array)?,
+        };
+
+        // iii. If index ≥ len, return NormalCompletion(undefined).
+        if agent[iterator].next_index >= len {
+            agent[iterator].array = None;
+            return Ok(create_iter_result_object(agent, Value::Undefined, true).into_value());
+        }
+
+        // iv. Let indexNumber be 𝔽(index).
+        let index = SmallInteger::try_from(agent[iterator].next_index).unwrap();
+        // viii. Set index to index + 1.
+        agent[iterator].next_index += 1;
+
+        let result = match agent[iterator].kind {
+            // v. If kind is key, then
+            ArrayIteratorKind::Key => {
+                // 1. Let result be indexNumber.
+                Value::Integer(index)
+            }
+            // 3. If kind is value, then
+            ArrayIteratorKind::Value => {
+                // 1. Let elementKey be ! ToString(indexNumber).
+                // 2. Let elementValue be ? Get(array, elementKey).
+                // a. Let result be elementValue.
+                match array {
+                    Object::Array(array) if agent[array].object_index.is_none() => {
+                        let idx = usize::try_from(index.into_i64()).unwrap();
+                        array.as_slice(agent)[idx].unwrap_or(Value::Undefined)
```

Comment:
**issue:** If the slot contains None or is out of bounds, then we need to go into the prototype chain or call possibly call a getter.

Checking the slot being Some can also be done even if we have a backing object.

The internal_get method of Array could check that for indexed properties directly actually.